chain-signatures/node/src/storage/presignature_storage.rs

```diff
@@ -231,31 +243,53 @@ impl PresignatureStorage {
         "#;
 
         let Some(mut conn) = self.connect().await else {
-            return Vec::new();
+            return R::default();
         };
-        let result: Result<Vec<PresignatureId>, _> = redis::Script::new(SCRIPT)
+        let result: Result<R, _> = redis::Script::new(SCRIPT)
             .key(&self.presig_key)
             .key(&self.reserved_key)
             .key(owner_key(&self.owner_keys, owner))
             // NOTE: this encodes each entry of owner_shares as a separate ARGV[index] entry.
-            .arg(owner_shares)
+            .arg(owner_shares.into_iter().copied().collect::<Vec<_>>())
             .invoke_async(&mut conn)
             .await;
 
         match result {
-            Ok(outdated) => {
-                if !outdated.is_empty() {
-                    tracing::info!(?outdated, "removed outdated presignatures");
-                }
-                outdated
-            }
+            Ok(outdated) => outdated,
             Err(err) => {
                 tracing::warn!(?err, "failed to remove outdated presignatures");
-                Vec::new()
+                R::default()
             }
         }
     }
 
+    /// Kicks participants from the given presignatures.
+    pub async fn kick_participants(&self, kick: HashMap<PresignatureId, Vec<Participant>>) {
+        if kick.is_empty() {
+            return;
+        }
+        let Some(mut conn) = self.connect().await else {
+            return;
+        };
+        let mut pipe = redis::pipe();
```

Comment:
We were not using pipelines before, since they are not atomic. We can face race conditions.

Reply:
look at the line below, calling `pipe.atomic()` makes it atomic

Reply:
You are sending all the actions in one batch, that is good. But during the execution they are not atomic.
It’s not terrible for this type of operation, though.

Reply:
I'm not sure about this. [`pipe.atomic()`](https://docs.rs/redis/latest/redis/struct.Pipeline.html#method.atomic) wraps the commands in `MULTI` + `EXEC` which creates a [redis transaction](https://redis.io/docs/latest/develop/interact/transactions/). As I understand this, it should be just as atomic as a script. But I'm not super familiar with these operations.

Reply:
Oh... I'm quitting AI tools. In fact, MULTI + EXEC is atomic. Works a bit differently. I think for us, the biggest difference is that scripts are blocking all other operations during execution. That is more predictable, but still, this solution should work perfectly.